libgit/browser.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	c, err := repo.CommitObject(ref.Hash())
+	if err != nil {
+		return nil, err
+	}
+	tree, err := c.Tree()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Browser{
+		tree:  tree,
+		root:  string(gitBranchName),
+		mtime: time.Now(),
+	}, nil
+}
+
+///// Read-only functions:
+
+const (
+	maxSymlinkLevels = 40 // same as Linux
+)
+
+func (b *Browser) readLink(filename string) (string, error) {
+	f, err := b.tree.File(filename)
+	if err != nil {
+		return "", err
+	}
+	r, err := f.Reader()
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func (b *Browser) followSymlink(filename string) (string, error) {
+	// Otherwise, resolve the symlink and return the underlying FileInfo.
+	link, err := b.readLink(filename)
+	if err != nil {
+		return "", err
+	}
+	if path.IsAbs(link) {
+		return "", errors.Errorf("Can't follow absolute link: %s", link)
+	}
+
+	parts := strings.Split(filename, "/")
+	var parentPath string
+	if len(parts) > 0 {
+		parentPath = path.Join(parts[:len(parts)-1]...)
+	}
+	newPath := path.Clean(path.Join(parentPath, link))
+	if strings.HasPrefix(newPath, "..") {
+		return "", errors.Errorf(
+			"Cannot follow symlink out of chroot: %s", newPath)
+	}
+	return newPath, nil
+}
+
+// Open implements the billy.Filesystem interface for Browser.
+func (b *Browser) Open(filename string) (billy.File, error) {
+	for i := 0; i < maxSymlinkLevels; i++ {
+		fi, err := b.Lstat(filename)
+		if err != nil {
+			return nil, err
+		}
+
+		// If it's not a symlink, we can return right away.
+		if fi.Mode()&os.ModeSymlink == 0 {
+			f, err := b.tree.File(filename)
+			if err != nil {
+				return nil, err
+			}
+			return newBrowserFile(f)
+		}
+
+		filename, err = b.followSymlink(filename)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, errors.New("Cannot resolve deep symlink chain")
+}
+
+// OpenFile implements the billy.Filesystem interface for Browser.
+func (b *Browser) OpenFile(filename string, flag int, _ os.FileMode) (
+	f billy.File, err error) {
+	if flag&os.O_CREATE != 0 {
+		return nil, errors.New("Browser can't create files")
+	}
+
+	return b.Open(filename)
+}
+
+// Lstat implements the billy.Filesystem interface for Browser.
+func (b *Browser) Lstat(filename string) (fi os.FileInfo, err error) {
+	entry, err := b.tree.FindEntry(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var size int64
+	if entry.Mode.IsFile() {
+		f, err := b.tree.File(filename)
+		if err != nil {
+			return nil, err
+		}
+		size = f.Size
+	} else {
+		// Estimate directory size by the number of entries.
+		dirTree, err := b.tree.Tree(filename)
+		if err != nil {
+			return nil, err
+		}
+		for _, e := range dirTree.Entries {
+			size += int64(len(e.Name))
+		}
+	}
+
+	// Git doesn't keep track of the mtime anywhere, so just use the
+	// time this browser was created.
+	return &browserFileInfo{entry, size, b.mtime}, nil
```

Comment:
Would it be better to use the commit's timestamp instead?

Reply:
Good idea, will do.